Feature/v0.9 upgrade

KwilLuke opened the pull request.

martin-opensky committed:
feat: deprecate legacy database APIs

Removed broadcast() method and deprecated database management APIs (deploy, drop, getDBID,
getSchema, listDatabases)

BREAKING CHANGE: Remove broadcast() method and deprecate database management APIs (deploy, drop,
getDBID, getSchema, listDatabases) in favor of executeSQL and selectQuery APIs

martin-opensky committed:
feat: deprecate legacy selectQuery signature

Deprecate selectQuery(dbid, query) signature in favor of selectQuery(query, params?, signer?) to
align with removal of DBID and addition of params & signer

BREAKING CHANGE: Deprecate selectQuery(dbid, query) signature. The new signature selectQuery(query,
params?, signer?) should be used instead

martin-opensky committed:
feat: deprecate DBID in call() requests and require namespace instead

The dbid field is no longer used in call() requests and has been marked as deprecated in favour of
the namespace field.  CallBody interface created to replace the ActionBody interface for call()
requests

BREAKING CHANGE: deprecate dbid field in call() actionBody.  Require namespace instead.

martin-opensky committed:
feat: add support for encoded params in selectQuery API

martin-opensky committed:
fix: remove params required in selectQuery signature / add dbid to deprecated queries

When the dbid is passed as a parameter it will be included in the query in order to make deprecated
queries fully functional

martin-opensky committed:
feat: add new validation for action inputs when calling action

This is a working progress.  Pushing changes so far.

martin-opensky committed:
feat: add completed call() action code / deprecate ActionInputs

View action code call() now integrated with new validation approach using the selectQuery() method
to get action definition.  ActionInputs deprecated and all inputs converted to Entries[]

martin-opensky committed:
refactor: improve parameter method names

renamed format to encode

martin-opensky committed:
chore: add examples directory to the gitignore

enabling the encoding examples to be saved locally for easy integration into the SDK

RawStatement and ActionExecution serialization examples

add remaining payloads, encoded value changes

Update call_payload.ts

change raw stmt example to encodeScalar

chore: remove martin_examples from my branch

martin-opensky committed:
feat: add call action encoding and selectQuery params encoding

General refactor, improved types and location of methods, interfaces, enums.  Call Action now uses
new Kwil encoding methods as does selectQuery for params encoding

martin-opensky committed:
feat: add support for exec action encoding

the new kwil encoding has been implemented in the SDK to replace the older rlp approach used by
kwild

martin-opensky committed:
feat: add new encoding for auth transfer

the new approach for kwild encoding has been integrated for the funder transfer api

martin-opensky committed:
feat: add new query API for exec queries

implemented new api using the kwild encoding

martin-opensky committed:
feat: update txInfo() response to reflect change in kwild (hash -> tx_hash)

The response data now returns tx_hash instead of hash

BREAKING CHANGE: Now when calling txInfo() the response data includes the transaction hash in the
'tx_hash' property instead of 'hash'

martin-opensky committed:
refactor: fix decimal metadata precision / scale

Fix ensures the decimal var type works as expected in kwild and postgres

martin-opensky committed:
test: upgrade tests to support new kwild changes

Implement tests that use new CREATE NAMESPACE, CREATE ACTION, CREATE TABLE SQL syntax and test these
actions.  Add the execSql and selectQuery API tests

martin-opensky committed:
chore: remove console.logs, refactor integration tests

martin-opensky committed:
refactor: refactor action input validation and encoding

action validation & input encoding is now shared between action builder and auth

martin-opensky committed:
chore: remove unneeded imports / update App.tsx

martin-opensky committed:
fix: fix the action call() signer when kwild in private mode

working progress but push to include changes so far

martin-opensky committed:
feat: add action input encoding for private kwild nodes / update encoding for authenticating call()

authenticatePrivateMode() now has new encoding for authenticating call() actions in private mode.
Signature updated for user.call.  Add support for action input encoding in private mode when action
schema is unavailable

martin-opensky committed:
test: add database integration and variable integration tests

martin-opensky committed:
refactor: remove comments / move namespace methods into another file / update integration tests

martin-opensky committed:
refactor: minor changes to integration tests

martin-opensky committed:
refactor: update action input encoding to make simpler. remove complex way of finding param type

martin-opensky committed:
refactor: add comments and improve parameterEncoding.ts method names

martin-opensky committed:
feat: add support for arrays in parameter encoding for queries and actions

martin-opensky committed:
feat: add array integration tests

martin-opensky committed:
refactor: test refactor

cleanup and tests

resolve review comments

KwilLuke commented on src/utils/keys.ts:
```diff
@@ -1,46 +1,73 @@
+  let keyType = AccountKeyType.SECP256K1;
+
+  if (typeof owner === 'string') {
+    if (isNearPubKey(owner)) {
```
We will need to change this. The Near pub key with `ed25519` prefix is not the same as the ed25519 keys that kwil supports.

KwilLuke commented on src/client/kwil.ts:
```diff
@@ -76,75 +78,30 @@ export abstract class Kwil<T extends EnvironmentType> extends Client {
-      owner = owner.toLowerCase();
-      owner = hexToBytes(owner);
-    }
+    const accountId = getAccountId(owner);
```
I don't think we should infer the account id. I think we should optionally let the user pass an account ID (in case they are using custom signers), and then attempt to infer the account ID if none is passed.

KwilLuke commented on src/core/enums.ts:
```diff
@@ -46,11 +43,13 @@ export enum EncodingType {
+  // TODO:  delete deprecated payload types
   DEPLOY_DATABASE = 'deploy_schema',
   DROP_DATABASE = 'drop_schema',
   EXECUTE_ACTION = 'execute',
```
Call action isn't a payload per se.

KwilLuke replied:
Disregarding for now. Will require substantial refactoring.

KwilLuke commented on src/core/network.ts:
```diff
@@ -1,30 +1,36 @@
+export interface AccountId {
+  identifier: string;
+  key_type: AccountKeyType;
+}
```
This needs support for custom signers.

KwilLuke commented on src/funder/funder.ts:
```diff
@@ -36,15 +37,10 @@ export class Funder<T extends EnvironmentType> {
-    } else {
-      to = payload.to;
-    }
+    const accountId = getAccountId(payload.to);
```
We need to let the user pass account id type (to support custom signers). By default, it can use secp256k1

KwilLuke commented on src/utils/keys.ts:
```diff
@@ -1,46 +1,73 @@
 export function isNearPubKey(pubKey: string): boolean {
-    return pubKey.startsWith('ed25519:');
+  return pubKey.startsWith('ed25519:');
 }
```
Let's get rid of this.

KwilLuke commented on src/utils/parameterEncoding.ts:
```diff
@@ -0,0 +1,175 @@
+  const dataType: DataType = {
+    name: varType,
+    is_array: Array.isArray(val),
+    ...metadataSpread,
```
Let's attach metadata every time. If it is not a decimal, the metadata will be [0,0]

KwilLuke reviewed (commented):
fixes needed

KwilLuke committed:
ci: update ci for kwil-db v0.10

fix ci

Update ci.yaml

set db-owner

Update ci.yaml

Update ci.yaml

Update ci.yaml

Update ci.yaml

Update ci.yaml

Update ci.yaml

KwilLuke committed:
docs(readme): update readme

KwilLuke commented:
Still need: Positional Parameters

KwilLuke committed:
feat: positional arguments

Support positional arguments in action execution and action call.

BREAKING CHANGE: For the `kwil.call` api, the `inputs` field only takes an object or a tuple of
scalar values. Previously, `inputs` received an array of only one object. We determined the single
length array was unncessary and therefore removed it.

fix: minor bug fixes

KwilLuke committed:
docs: migration guide

add migration guide

Update ci.yaml

update doocs

update docs

Update ci.yaml

Update README.md

brennanjl commented on README.md:
```diff
@@ -161,59 +125,51 @@ const res = await kwil.execute(actionBody, kwilSigner)
+        inputs: [
+            ['input_value_1', 'input_value_2', 'input_value_3'],
+            ['input_value_1', 'input_value_2', 'input_value_3'],
+            ['input_value_1', 'input_value_2', 'input_value_3']
```
Seems a bit confusing that you have the same input value 3 times over. Could you make it clear that they're supposed to be different iterations?

```js
['iter_1_value_1', 'iter_1_value_2', 'iter_1_value_3'],
['iter_2_value_1', 'iter_2_value_2', 'iter_2_value_3'],
```

brennanjl replied:
otherwise lgtm

KwilLuke committed:
feat(index.ts): update type exports

All relevant types for public kwil-js methods are exported in the `Types` namespace.
